# 1. Conociendo el eventing data y los archivos JSON

Es hora de adentrarnos en el análisis del eventing data, que son todos los datos recogidos, relacionados con lo que sucede con la pelota. Quién pasa el balón, desde donde lo pasa, qué tipo de pase, donde finaliza el pase, quien recibe el pase, desde donde se tira, en qué acabó el tiro, etc. Toda esta información y muchas más es la que se conoce como eventing data. 

Este tipo de data está guardada en archivos de formato tipo JSON, vimos el archivo tipo CSV que almacenamos la información en un texto plano separada por comas. Con el formato JSON, la cosa cambia, utilizamos diccionarios para almacenar la información. Pasemos a adentrarnos en este apasionante mundo del análisis de datos con el formato JSON. 


In [ ]:
import pandas as pd
import json

# 1.1. Leer un archivo JSON
Utilizamos la palabra reservada del lenguaje with más la función open() para leer un archivo JSON. De esta manera podremos ver la información que contiene el archivo de datos competitions.json

In [ ]:
with open('/content/drive/MyDrive/Mi curso/Data/event_data/Statsbomb/data/competitions.json') as f:
  competitions = json.load(f)

# 1.2. Estructura interna de una archivo JSON
Una vez importado el archivo con la función json.load() veamos la estructura interna del archivo JSON

In [ ]:
competitions

[{'competition_gender': 'male',
  'competition_id': 16,
  'competition_name': 'Champions League',
  'country_name': 'Europe',
  'match_available': '2020-10-25T12:33:27.855343',
  'match_updated': '2020-10-25T12:33:27.855343',
  'season_id': 4,
  'season_name': '2018/2019'},
 {'competition_gender': 'male',
  'competition_id': 16,
  'competition_name': 'Champions League',
  'country_name': 'Europe',
  'match_available': '2020-07-29T05:00',
  'match_updated': '2020-07-29T05:00',
  'season_id': 1,
  'season_name': '2017/2018'},
 {'competition_gender': 'male',
  'competition_id': 16,
  'competition_name': 'Champions League',
  'country_name': 'Europe',
  'match_available': '2020-07-29T05:00',
  'match_updated': '2020-08-26T12:33:15.869622',
  'season_id': 2,
  'season_name': '2016/2017'},
 {'competition_gender': 'male',
  'competition_id': 16,
  'competition_name': 'Champions League',
  'country_name': 'Europe',
  'match_available': '2020-07-29T05:00',
  'match_updated': '2020-08-26T12:33:1

Esto es solo un extracto de la información que contiene, pero como podemos ver no es más que una estructura de diccionarios. Podemos utilizar la función len() para saber la cantidad de diccionarios dentro de nuestro archivo JSON.

In [ ]:
len(competitions)

37

Como hemos visto, los diccionarios se encuentran dentro de una lista, por tanto podemos acceder a la información gracias a la técnica de los índices de las listas, que ya hemos ido viendo durante todo el curso. Vamos a Acceder a los valores del primer elemento del archivo JSON.

In [ ]:
competitions[0]

{'competition_gender': 'male',
 'competition_id': 16,
 'competition_name': 'Champions League',
 'country_name': 'Europe',
 'match_available': '2020-10-25T12:33:27.855343',
 'match_updated': '2020-10-25T12:33:27.855343',
 'season_id': 4,
 'season_name': '2018/2019'}

Ahora con lo aprendido en el curso, podemos ir accediendo a la información, ya que estamos obteniendo un diccionario y nosotros ya sabemos como acceder a la llaves de un diccionario.

In [ ]:
competitions[0]['competition_name']

'Champions League'

Vamos a iterar sobre el valor de todas las llaves de 'competition_name' de cada diccionario.

In [ ]:
for competition in competitions:
  print(competition['competition_name'])

Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
Champions League
FA Women's Super League
FA Women's Super League
FIFA World Cup
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
La Liga
NWSL
Premier League
Women's World Cup


Creamos una lista vacía, para a través de un bucle for, añadir las diferentes competiciones.  

In [ ]:
competition_name = []

for competition in competitions:
  if competition['competition_name'] not in competition_name:
    competition_name.append(competition['competition_name'])

competition_name

['Champions League',
 "FA Women's Super League",
 'FIFA World Cup',
 'La Liga',
 'NWSL',
 'Premier League',
 "Women's World Cup"]

Realizamos el mismo procedimiento para saber el id de cada competición.

In [ ]:
competition_name = []
competition_id = []

for competition in competitions:
  if competition['competition_name'] not in competition_name:
    competition_name.append(competition['competition_name'])

  if competition['competition_id'] not in competition_id:
    competition_id.append(competition['competition_id'])


In [ ]:
competition_id

[16, 37, 43, 11, 49, 2, 72]

Utilizamos la función list() y zip() para juntar los valores de los id de las competiciones y el nombre de las competiciones.

In [ ]:
list(zip(competition_id, competition_name))

[(16, 'Champions League'),
 (37, "FA Women's Super League"),
 (43, 'FIFA World Cup'),
 (11, 'La Liga'),
 (49, 'NWSL'),
 (2, 'Premier League'),
 (72, "Women's World Cup")]

Realizamos el mismo procedimiento para extraer el id de la temporada y el año de la temporada.

In [ ]:
season_id = []
season_name = []

for season in competitions:
        if season['season_id'] not in season_id:
            season_id.append(season['season_id'])
        if season['season_name'] not in season_name:
            season_name.append(season['season_name'])

In [ ]:
list(zip(season_id, season_name))

[(4, '2018/2019'),
 (1, '2017/2018'),
 (2, '2016/2017'),
 (27, '2015/2016'),
 (26, '2014/2015'),
 (25, '2013/2014'),
 (24, '2012/2013'),
 (23, '2011/2012'),
 (22, '2010/2011'),
 (21, '2009/2010'),
 (41, '2008/2009'),
 (39, '2006/2007'),
 (37, '2004/2005'),
 (44, '2003/2004'),
 (76, '1999/2000'),
 (42, '2019/2020'),
 (3, '2018'),
 (40, '2007/2008'),
 (38, '2005/2006'),
 (30, '2019')]

Del archivo competitions.json hemos extraído las diferentes competiciones de las que tenemos datos con su respectivo id y también las diferentes temporadas disponibles con su respectivo id.

# 1.3. Acceder a la información de los partidos 
Una vez entendido cómo están estructurados los datos a través de las carpetas y cómo necesitamos saber los id, para acceder a los diferentes partidos. Vamos a acceder a la carpeta matches, seguidamente acceder a la carpeta con el id 16 que corresponde a los datos de la Champions League y una vez dentro, seleccionar el archivo 4.json el cual  id 4, corresponde a la temporada 2018/2019.
Accedemos a ese archivo JSON de la misma manera que hicimos con el archivo competitions.json, solo necesitamos pasarle la ruta a la función open()

In [ ]:
with open('/content/drive/MyDrive/Mi curso/Data/event_data/Statsbomb/data/matches/16/4.json') as f:
  matches = json.load(f)

Ahora tenemos la información de los partidos en formato JSON, de la Champions League de la temporada 2018/2019

In [ ]:
matches

[{'away_score': 2,
  'away_team': {'away_team_gender': 'male',
   'away_team_group': None,
   'away_team_id': 24,
   'away_team_name': 'Liverpool',
   'country': {'id': 68, 'name': 'England'},
   'managers': [{'country': {'id': 85, 'name': 'Germany'},
     'dob': None,
     'id': 94,
     'name': 'Jürgen Klopp',
     'nickname': None}]},
  'competition': {'competition_id': 16,
   'competition_name': 'Champions League',
   'country_name': 'Europe'},
  'competition_stage': {'id': 26, 'name': 'Final'},
  'home_score': 0,
  'home_team': {'country': {'id': 68, 'name': 'England'},
   'home_team_gender': 'male',
   'home_team_group': None,
   'home_team_id': 38,
   'home_team_name': 'Tottenham Hotspur',
   'managers': [{'country': {'id': 11, 'name': 'Argentina'},
     'dob': None,
     'id': 81,
     'name': 'Mauricio Roberto Pochettino Trossero',
     'nickname': 'Mauricio Pochettino'}]},
  'kick_off': '21:00:00.000',
  'last_updated': '2020-07-29T05:00',
  'match_date': '2019-06-01',
  'mat

Si utilizamos len(), veremos que este archivo solo contiene un diccionario, por tanto, sólo contiene un partido.

In [ ]:
len(matches)

1

Las llaves del diccionario nos dirá el tipo de información que tiene el archivo JSON. Podemos acceder a las llaves a través de .Keys()

In [ ]:
matches[0].keys()

dict_keys(['match_id', 'match_date', 'kick_off', 'competition', 'season', 'home_team', 'away_team', 'home_score', 'away_score', 'match_status', 'last_updated', 'metadata', 'match_week', 'competition_stage'])

Con la llave 'home_team' accederemos a la información del equipo que juega en casa. Como veis a la vez tiene diccionarios en su interior.

In [ ]:
matches[0]['home_team']

{'country': {'id': 68, 'name': 'England'},
 'home_team_gender': 'male',
 'home_team_group': None,
 'home_team_id': 38,
 'home_team_name': 'Tottenham Hotspur',
 'managers': [{'country': {'id': 11, 'name': 'Argentina'},
   'dob': None,
   'id': 81,
   'name': 'Mauricio Roberto Pochettino Trossero',
   'nickname': 'Mauricio Pochettino'}]}

Accedemos al nombre del equipo a través de múltiples llaves.

In [ ]:
matches[0]['home_team']['home_team_name']

'Tottenham Hotspur'

Hacemos exactamente lo mismo con el equipo visitante.

In [ ]:
matches[0]['away_team']['away_team_name']

'Liverpool'

Accedemos al marcador del equipo visitante.

In [ ]:
matches[0]['away_score']

2

Accedemos al marcador del equipo que juega en casa.

In [ ]:
matches[0]['home_score']

0

Accedemos al nombre de la fase de la competición, en este caso la final.

In [ ]:
matches[0]['competition_stage']['name']

'Final'

Fijaros como en la carpeta de matches, se encuentran la información de los partidos, ordenados por temporada y competición. En estos archivos podemos ver la información básica de un partido. Accedemos ahora a los partidos de La Liga, que corresponde a la carpeta con el Id 4.

In [ ]:
with open('/content/drive/MyDrive/Mi curso/Data/event_data/Statsbomb/data/matches/11/4.json') as f:
  matches = json.load(f)

Al realizar len(), vemos que obtenemos 34 diccionarios, por tanto tenemos 34 partidos.

In [ ]:
len(matches)

34

Si accedemos al primero de ellos podemos ver de qué partidos se trata. 

In [ ]:
matches[0]

{'away_score': 2,
 'away_team': {'away_team_gender': 'male',
  'away_team_group': None,
  'away_team_id': 207,
  'away_team_name': 'Valencia',
  'country': {'id': 214, 'name': 'Spain'},
  'managers': [{'country': {'id': 214, 'name': 'Spain'},
    'dob': None,
    'id': 211,
    'name': 'Marcelino García Toral',
    'nickname': 'Marcelino'}]},
 'competition': {'competition_id': 11,
  'competition_name': 'La Liga',
  'country_name': 'Spain'},
 'competition_stage': {'id': 1, 'name': 'Regular Season'},
 'home_score': 2,
 'home_team': {'country': {'id': 214, 'name': 'Spain'},
  'home_team_gender': 'male',
  'home_team_group': None,
  'home_team_id': 217,
  'home_team_name': 'Barcelona',
  'managers': [{'country': {'id': 214, 'name': 'Spain'},
    'dob': None,
    'id': 227,
    'name': 'Ernesto Valverde Tejedor',
    'nickname': 'Ernesto Valverde'}]},
 'kick_off': '18:30:00.000',
 'last_updated': '2020-07-29T05:00',
 'match_date': '2019-02-02',
 'match_id': 16157,
 'match_status': 'availabl

Nuevamente accedemos a la información a través de las llaves del diccionario, como hicimos con los datos de la Champions League.

In [ ]:
print(matches[0]['home_team']['home_team_name'])
print(matches[0]['away_team']['away_team_name'])

Barcelona
Valencia


Accedemos al resultado del partido.

In [ ]:
print(matches[0]['home_score'])
print(matches[0]['away_score'])

2
2


Accedemos al tipo de partido, en este caso es un partido de temporada regular. 

In [ ]:
matches[0]['competition_stage']['name']

'Regular Season'

# 1.4. Extraer la información de los partidos 
Vamos a realizar un bucle for, donde extraemos todos los nombres de los equipos, el resultado y el Id del partido. Para realizar un print y poder ver de una mejor manera los partidos y resultados que tenemos en este archivo JSON.


In [ ]:
for match in matches:
  home_team_name = match['home_team']['home_team_name']
  away_team_name = match['away_team']['away_team_name']
  home_score = match['home_score']
  away_score = match['away_score']
  match_id = match['match_id']

  print(home_team_name, 'vs', away_team_name, ':', str(home_score), '-', str(away_score), '|', str(match_id))


Barcelona vs Valencia : 2 - 2 | 16157
Barcelona vs Leganés : 3 - 1 | 16136
Real Betis vs Barcelona : 1 - 4 | 16215
Barcelona vs Atlético Madrid : 2 - 0 | 16248
Barcelona vs Real Betis : 3 - 4 | 16056
Barcelona vs Levante : 1 - 0 | 16289
Barcelona vs Villarreal : 2 - 0 | 16079
Girona vs Barcelona : 0 - 2 | 16149
Deportivo Alavés vs Barcelona : 0 - 2 | 16275
Barcelona vs Huesca : 8 - 2 | 15973
Eibar vs Barcelona : 2 - 2 | 16317
Barcelona vs Eibar : 3 - 0 | 16131
Barcelona vs Celta Vigo : 2 - 0 | 16109
Barcelona vs Rayo Vallecano : 3 - 1 | 16205
Levante vs Barcelona : 0 - 5 | 16095
Barcelona vs Getafe : 2 - 0 | 16306
Barcelona vs Athletic Bilbao : 1 - 1 | 16010
Barcelona vs Girona : 2 - 2 | 15986
Villarreal vs Barcelona : 4 - 4 | 16240
Real Sociedad vs Barcelona : 1 - 2 | 15978
Athletic Bilbao vs Barcelona : 0 - 0 | 16173
Valencia vs Barcelona : 1 - 1 | 16023
Barcelona vs Real Sociedad : 2 - 1 | 16265
Real Madrid vs Barcelona : 0 - 1 | 16196
Real Valladolid vs Barcelona : 0 - 1 | 15956
Ba

Ahora tenemos la información de los partidos y de su ID, para entonces poder ir a buscar en la carpeta events los eventos de dicho partido. Vamos a seleccionar el partido del Betis vs Barcelona que terminó en 1-4. Lo que necesito para acceder a los eventos de ese partido es el ID del partido. 

## 1.5. Acceder al eventing data de un partido 
Gracias a todo el proceso anterior, ahora podemos acceder al partido que a nosotros nos interese analizar su eventing data. En la carpeta events tenemos cientos de partidos para analizar, pero todo el proceso anterior, ha sido necesario para saber cómo encontrar el partido que nos interese.

In [ ]:
with open('/content/drive/MyDrive/Mi curso/Data/event_data/Statsbomb/data/events/16215.json', encoding='utf8') as f:
  data = json.load(f)

Con los archivos JSON que contienen los eventos del partido, con len() podemos observar la cantidad de eventos.

In [ ]:
len(data)

3998

Disponemos de 3998 eventos, cada diccionario dentro del archivo JSON corresponde a un evento.

# 1.6. Acceder a cada uno de los eventos 
Si vamos accediendo a través de índices, iremos viendo los diferentes eventos recogidos. El primer evento es la alineación inicial del Betis.


In [ ]:
data[0]

{'duration': 0.0,
 'id': '726dfc28-a2ee-446f-80e7-7cdb6c807b75',
 'index': 1,
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 1, 'name': 'Regular Play'},
 'possession': 1,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'second': 0,
 'tactics': {'formation': 352,
  'lineup': [{'jersey_number': 13,
    'player': {'id': 6792, 'name': 'Pau López Sabata'},
    'position': {'id': 1, 'name': 'Goalkeeper'}},
   {'jersey_number': 23,
    'player': {'id': 6648, 'name': 'Aïssa Mandi'},
    'position': {'id': 3, 'name': 'Right Center Back'}},
   {'jersey_number': 5,
    'player': {'id': 7068, 'name': 'Marc Bartra Aregall'},
    'position': {'id': 4, 'name': 'Center Back'}},
   {'jersey_number': 12,
    'player': {'id': 6777, 'name': 'Sidnei Rechel da Silva Júnior'},
    'position': {'id': 5, 'name': 'Left Center Back'}},
   {'jersey_number': 17,
    'player': {'id': 6651, 'name': 'Joaquín Sánchez Rodríguez'},
    'position': {'id': 7, 'name': 'Right Wing Back'}},
   {'jersey_number': 

El segundo elemento es la alineación inicial del Barcelona.

In [ ]:
data[1]

{'duration': 0.0,
 'id': '4144e8d8-3f1b-4600-8729-8d75d61451ef',
 'index': 2,
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 1, 'name': 'Regular Play'},
 'possession': 1,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'second': 0,
 'tactics': {'formation': 442,
  'lineup': [{'jersey_number': 1,
    'player': {'id': 20055, 'name': 'Marc-André ter Stegen'},
    'position': {'id': 1, 'name': 'Goalkeeper'}},
   {'jersey_number': 20,
    'player': {'id': 6379, 'name': 'Sergi Roberto Carnicer'},
    'position': {'id': 2, 'name': 'Right Back'}},
   {'jersey_number': 3,
    'player': {'id': 5213, 'name': 'Gerard Piqué Bernabéu'},
    'position': {'id': 3, 'name': 'Right Center Back'}},
   {'jersey_number': 15,
    'player': {'id': 6826, 'name': 'Clément Lenglet'},
    'position': {'id': 5, 'name': 'Left Center Back'}},
   {'jersey_number': 18,
    'player': {'id': 5211, 'name': 'Jordi Alba Ramos'},
    'position': {'id': 6, 'name': 'Left Back'}},
   {'jersey_number': 4,
    'play

Si nos fijamos, los eventos tienen la llave 'type' que nos informa del tipo de evento. En ambos casos corresponde al evero 'Starting XI' que es el 11 inicial.

Los eventos del diccionario 3 y 4 corresponden simplemente al inicio de la primera parte para ambos equipos.


In [ ]:
data[2]

{'duration': 0.0,
 'id': 'fab0e530-5c44-47ed-b963-e454c2321cfe',
 'index': 3,
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 1, 'name': 'Regular Play'},
 'possession': 1,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['961f1ffa-943d-4674-8e31-accf91c8cf6c'],
 'second': 0,
 'team': {'id': 218, 'name': 'Real Betis'},
 'timestamp': '00:00:00.000',
 'type': {'id': 18, 'name': 'Half Start'}}

In [ ]:
data[3]

{'duration': 0.0,
 'id': '961f1ffa-943d-4674-8e31-accf91c8cf6c',
 'index': 4,
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 1, 'name': 'Regular Play'},
 'possession': 1,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['fab0e530-5c44-47ed-b963-e454c2321cfe'],
 'second': 0,
 'team': {'id': 217, 'name': 'Barcelona'},
 'timestamp': '00:00:00.000',
 'type': {'id': 18, 'name': 'Half Start'}}

Si nos fijamos en 'type' vemos que es el evento 'Half Start'. Es a partir del evento 5 donde se empiezan a recoger los eventos que suceden con la pelota. 

 

In [ ]:
data[4]

{'duration': 0.508755,
 'id': '63a0b71c-9d24-432b-90d9-bcab133ec115',
 'index': 5,
 'location': [60.0, 40.0],
 'minute': 0,
 'pass': {'angle': -2.819842,
  'body_part': {'id': 40, 'name': 'Right Foot'},
  'end_location': [54.0, 38.0],
  'height': {'id': 1, 'name': 'Ground Pass'},
  'length': 6.3245554,
  'recipient': {'id': 5563, 'name': 'José Andrés Guardado Hernández'},
  'type': {'id': 65, 'name': 'Kick Off'}},
 'period': 1,
 'play_pattern': {'id': 9, 'name': 'From Kick Off'},
 'player': {'id': 3990, 'name': 'Jesé Rodríguez Ruiz'},
 'position': {'id': 24, 'name': 'Left Center Forward'},
 'possession': 2,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['5bad9e2c-feec-40ad-96db-b4c76dea8545'],
 'second': 0,
 'team': {'id': 218, 'name': 'Real Betis'},
 'timestamp': '00:00:00.697',
 'type': {'id': 30, 'name': 'Pass'}}

Aquí ya podemos ver mucha más información:
- 'location': Las coordenadas donde se produce el pase
- 'player': Quien realiza el pase
- 'type': El tipo de evento, en este caso 'Pass'
- 'pass': Al ser el evento pase, obtener información del tipo de pase y sus características 
  - 'end_location': Las coordenadas donde finaliza el pase
  - 'body_part': La parte del cuerpo con la que se hizo el pase
  - 'recipient’: Quien recibe el pase


El siguiente evento, hace referencia a la recepción del pase anterior.


In [ ]:
data[5]

{'id': '5bad9e2c-feec-40ad-96db-b4c76dea8545',
 'index': 6,
 'location': [54.0, 38.0],
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 9, 'name': 'From Kick Off'},
 'player': {'id': 5563, 'name': 'José Andrés Guardado Hernández'},
 'position': {'id': 15, 'name': 'Left Center Midfield'},
 'possession': 2,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['63a0b71c-9d24-432b-90d9-bcab133ec115'],
 'second': 1,
 'team': {'id': 218, 'name': 'Real Betis'},
 'timestamp': '00:00:01.206',
 'type': {'id': 42, 'name': 'Ball Receipt*'}}

Una vez recibido el pase, se recoge lo que hace el jugador con la pelota. En este caso es un evento de tipo 'Carry' que se refiere a que se desplaza con la pelota.

In [ ]:
data[6]

{'carry': {'end_location': [56.0, 33.0]},
 'duration': 1.963745,
 'id': '5d34f03b-9fdd-48c4-bcb2-f1fbcb4fc045',
 'index': 7,
 'location': [54.0, 38.0],
 'minute': 0,
 'period': 1,
 'play_pattern': {'id': 9, 'name': 'From Kick Off'},
 'player': {'id': 5563, 'name': 'José Andrés Guardado Hernández'},
 'position': {'id': 15, 'name': 'Left Center Midfield'},
 'possession': 2,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['5bad9e2c-feec-40ad-96db-b4c76dea8545',
  '7fe2a2af-eaae-4828-8ffb-cc950067a400'],
 'second': 1,
 'team': {'id': 218, 'name': 'Real Betis'},
 'timestamp': '00:00:01.206',
 'type': {'id': 43, 'name': 'Carry'}}

Después del 'Carry' realiza un pase y nuevamente obtendremos la información del pase. Desde donde se hace, quien lo recibe y características del pase.

In [ ]:
data[7]

{'duration': 1.557132,
 'id': '7fe2a2af-eaae-4828-8ffb-cc950067a400',
 'index': 8,
 'location': [56.0, 33.0],
 'minute': 0,
 'pass': {'angle': -2.6135182,
  'body_part': {'id': 40, 'name': 'Right Foot'},
  'end_location': [44.0, 26.0],
  'height': {'id': 1, 'name': 'Ground Pass'},
  'length': 13.892444,
  'recipient': {'id': 6777, 'name': 'Sidnei Rechel da Silva Júnior'}},
 'period': 1,
 'play_pattern': {'id': 9, 'name': 'From Kick Off'},
 'player': {'id': 5563, 'name': 'José Andrés Guardado Hernández'},
 'position': {'id': 15, 'name': 'Left Center Midfield'},
 'possession': 2,
 'possession_team': {'id': 218, 'name': 'Real Betis'},
 'related_events': ['96f5cc5a-41cf-4224-a00c-1fc4947c3c89'],
 'second': 3,
 'team': {'id': 218, 'name': 'Real Betis'},
 'timestamp': '00:00:03.170',
 'type': {'id': 30, 'name': 'Pass'}}

# 1.7. Convertir los archivos JSON a formato dataframe
Una vez entendida la estructura de nuestros  datos en formato JSON y cómo podemos acceder a la información que contienen, tenemos una función en pandas que nos permite convertir un archivo JSON a dataframe.


In [ ]:
df = pd.json_normalize(data, sep='_')
df.head()

,id,index,period,timestamp,minute,second,possession,duration,type_id,type_name,possession_team_id,possession_team_name,play_pattern_id,play_pattern_name,team_id,team_name,tactics_formation,tactics_lineup,related_events,location,player_id,player_name,position_id,position_name,pass_recipient_id,pass_recipient_name,pass_length,pass_angle,pass_height_id,pass_height_name,pass_end_location,pass_body_part_id,pass_body_part_name,pass_type_id,pass_type_name,carry_end_location,under_pressure,pass_outcome_id,pass_outcome_name,ball_receipt_outcome_id,...,goalkeeper_technique_name,goalkeeper_position_id,goalkeeper_position_name,goalkeeper_body_part_id,goalkeeper_body_part_name,goalkeeper_type_id,goalkeeper_type_name,goalkeeper_outcome_id,goalkeeper_outcome_name,clearance_aerial_won,shot_first_time,shot_saved_off_target,goalkeeper_shot_saved_off_target,bad_behaviour_card_id,bad_behaviour_card_name,pass_deflected,goalkeeper_end_location,pass_inswinging,pass_cut_back,foul_committed_type_id,foul_committed_type_name,injury_stoppage_in_chain,foul_won_defensive,pass_outswinging,shot_aerial_won,pass_goal_assist,clearance_left_foot,foul_committed_card_id,foul_committed_card_name,clearance_other,shot_one_on_one,shot_open_goal,substitution_outcome_id,substitution_outcome_name,substitution_replacement_id,substitution_replacement_name,foul_committed_offensive,pass_no_touch,dribble_overrun,dribble_no_touch
0,726dfc28-a2ee-446f-80e7-7cdb6c807b75,1,1,00:00:00.000,0,0,1,0.000000,35,Starting XI,218,Real Betis,1,Regular Play,218,Real Betis,352.0,"[{'player': {'id': 6792, 'name': 'Pau López Sa...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,4144e8d8-3f1b-4600-8729-8d75d61451ef,2,1,00:00:00.000,0,0,1,0.000000,35,Starting XI,218,Real Betis,1,Regular Play,217,Barcelona,442.0,"[{'player': {'id': 20055, 'name': 'Marc-André ...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,fab0e530-5c44-47ed-b963-e454c2321cfe,3,1,00:00:00.000,0,0,1,0.000000,18,Half Start,218,Real Betis,1,Regular Play,218,Real Betis,NaN,NaN,[961f1ffa-943d-4674-8e31-accf91c8cf6c],NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,961f1ffa-943d-4674-8e31-accf91c8cf6c,4,1,00:00:00.000,0,0,1,0.000000,18,Half Start,218,Real Betis,1,Regular Play,217,Barcelona,NaN,NaN,[fab0e530-5c44-47ed-b963-e454c2321cfe],NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,63a0b71c-9d24-432b-90d9-bcab133ec115,5,1,00:00:00.697,0,0,2,0.508755,30,Pass,218,Real Betis,9,From Kick Off,218,Real Betis,NaN,NaN,[5bad9e2c-feec-40ad-96db-b4c76dea8545],"[60.0, 40.0]",3990.0,Jesé Rodríguez Ruiz,24.0,Left Center Forward,5563.0,José Andrés Guardado Hernández,6.324555,-2.819842,1.0,Ground Pass,"[54.0, 38.0]",40.0,Right Foot,65.0,Kick Off,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Si utilizamos .shape podremos ver el tamaño del dataframe.

In [ ]:
df.shape

(3998, 121)

La columna type_name contiene los diferentes eventos dentro del JSON, con .unique()  podremos saber la cantidad de eventos que podemos analizar de un partido.

In [ ]:
df.type_name.unique()

array(['Starting XI', 'Half Start', 'Pass', 'Ball Receipt*', 'Carry',
       'Pressure', 'Clearance', 'Dribble', 'Duel', 'Block',
       'Ball Recovery', 'Interception', 'Dispossessed', 'Foul Committed',
       'Foul Won', 'Miscontrol', 'Shot', 'Goal Keeper', 'Dribbled Past',
       'Bad Behaviour', 'Injury Stoppage', 'Referee Ball-Drop',
       'Half End', 'Error', 'Substitution', 'Tactical Shift'],
      dtype=object)

## 1.7.1. Evaluamos el evento tiro
Filtramos los datos en base al evento 'Shot', para analizar la información de los tiros


In [ ]:
shots = df[df.type_name == 'Shot']
shots.head()

,id,index,period,timestamp,minute,second,possession,duration,type_id,type_name,possession_team_id,possession_team_name,play_pattern_id,play_pattern_name,team_id,team_name,tactics_formation,tactics_lineup,related_events,location,player_id,player_name,position_id,position_name,pass_recipient_id,pass_recipient_name,pass_length,pass_angle,pass_height_id,pass_height_name,pass_end_location,pass_body_part_id,pass_body_part_name,pass_type_id,pass_type_name,carry_end_location,under_pressure,pass_outcome_id,pass_outcome_name,ball_receipt_outcome_id,...,goalkeeper_technique_name,goalkeeper_position_id,goalkeeper_position_name,goalkeeper_body_part_id,goalkeeper_body_part_name,goalkeeper_type_id,goalkeeper_type_name,goalkeeper_outcome_id,goalkeeper_outcome_name,clearance_aerial_won,shot_first_time,shot_saved_off_target,goalkeeper_shot_saved_off_target,bad_behaviour_card_id,bad_behaviour_card_name,pass_deflected,goalkeeper_end_location,pass_inswinging,pass_cut_back,foul_committed_type_id,foul_committed_type_name,injury_stoppage_in_chain,foul_won_defensive,pass_outswinging,shot_aerial_won,pass_goal_assist,clearance_left_foot,foul_committed_card_id,foul_committed_card_name,clearance_other,shot_one_on_one,shot_open_goal,substitution_outcome_id,substitution_outcome_name,substitution_replacement_id,substitution_replacement_name,foul_committed_offensive,pass_no_touch,dribble_overrun,dribble_no_touch
586,6c30cac4-6f3b-4cd0-9913-26137b058416,587,1,00:10:56.243,10,56,25,1.179654,16,Shot,218,Real Betis,4,From Throw In,218,Real Betis,NaN,NaN,[4f0d23bf-492f-4f69-a394-f0c8eb847916],"[101.0, 22.0]",6658.0,Cristian Tello Herrera,8.0,Left Wing Back,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
647,67cfa552-952c-4f72-a33a-0dd6af217a66,648,1,00:12:18.100,12,18,30,1.226969,16,Shot,218,Real Betis,4,From Throw In,218,Real Betis,NaN,NaN,[59cdcec0-f1c7-4ba9-985f-21748bd54475],"[113.7, 39.9]",3990.0,Jesé Rodríguez Ruiz,24.0,Left Center Forward,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,True,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
834,8c95e369-ec54-4a84-9d3e-51abb6f38e6e,835,1,00:17:13.830,17,13,37,1.046942,16,Shot,217,Barcelona,3,From Free Kick,217,Barcelona,NaN,NaN,[0b60e19e-2902-49c4-aad0-05624d981cfd],"[95.6, 44.4]",5503.0,Lionel Andrés Messi Cuccittini,24.0,Left Center Forward,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1173,2ae506b9-0601-460b-b91d-802296120161,1174,1,00:24:28.346,24,28,51,1.375915,16,Shot,217,Barcelona,3,From Free Kick,217,Barcelona,NaN,NaN,[9ce51b00-00ed-4e9e-8e85-c1603569ae2e],"[94.0, 46.1]",5503.0,Lionel Andrés Messi Cuccittini,24.0,Left Center Forward,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1343,10dc9818-a8d2-47f9-b219-9bc87d7b22e6,1344,1,00:27:01.586,27,1,58,0.116586,16,Shot,218,Real Betis,3,From Free Kick,218,Real Betis,NaN,NaN,"[4c025aa6-1469-48c8-b228-5cc9dc1c542f, fdf6b77...","[104.1, 22.2]",6658.0,Cristian Tello Herrera,8.0,Left Wing Back,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN



Al realizar .shape vemos la cantidad de tiros que hubo en el partido.

In [ ]:
shots.shape

(29, 121)

Con shot_outcome_name accedemos al resultado del disparo.

In [ ]:
shots.shot_outcome_name

586                Saved
647     Saved Off Target
834                 Goal
1173               Saved
1343             Blocked
1368             Blocked
1393             Blocked
1417             Blocked
1645               Off T
1799               Saved
1824             Blocked
1871             Blocked
1963             Blocked
2022                Goal
2087               Off T
2207             Blocked
2377             Blocked
2675             Wayward
2740               Saved
2748               Off T
2816                Goal
3297               Saved
3420               Off T
3533             Blocked
3662                Goal
3700               Saved
3719                Goal
3852               Saved
3994                Post
Name: shot_outcome_name, dtype: object

Podemos realizar un .value_counts() para ver la cantidad de resultados obtenidos de un tiro.

In [ ]:
shots.shot_outcome_name.value_counts()

Blocked             10
Saved                7
Goal                 5
Off T                4
Post                 1
Wayward              1
Saved Off Target     1
Name: shot_outcome_name, dtype: int64

Con la columna 'shot_type_name' vemos si el tiro fue de un jugador abierto o desde falta.

In [ ]:
shots['shot_type_name'].value_counts()

Open Play    26
Free Kick     3
Name: shot_type_name, dtype: int64

Con posession_team_name sabremos a quién corresponde cada tiro, porque nos informa de la posesión del equipo en dicho evento.  

In [ ]:
shots.possession_team_name

586     Real Betis
647     Real Betis
834      Barcelona
1173     Barcelona
1343    Real Betis
1368    Real Betis
1393    Real Betis
1417    Real Betis
1645    Real Betis
1799     Barcelona
1824    Real Betis
1871     Barcelona
1963     Barcelona
2022     Barcelona
2087     Barcelona
2207    Real Betis
2377    Real Betis
2675     Barcelona
2740    Real Betis
2748     Barcelona
2816     Barcelona
3297     Barcelona
3420     Barcelona
3533    Real Betis
3662    Real Betis
3700     Barcelona
3719     Barcelona
3852     Barcelona
3994     Barcelona
Name: possession_team_name, dtype: object

Con .value_counts() podremos saber la cantidad de tiros de cada equipo.

In [ ]:
shots.possession_team_name.value_counts()

Barcelona     16
Real Betis    13
Name: possession_team_name, dtype: int64

Con la columna shots_statsbomb_xg vermos el valor de xG de cada tiro.

In [ ]:
shots.shot_statsbomb_xg

586     0.018497
647     0.388249
834     0.147725
1173    0.056689
1343    0.018782
1368    0.157381
1393    0.017350
1417    0.015063
1645    0.026171
1799    0.314474
1824    0.081565
1871    0.091432
1963    0.066856
2022    0.401625
2087    0.052468
2207    0.026238
2377    0.043450
2675    0.520617
2740    0.050398
2748    0.257788
2816    0.480570
3297    0.074971
3420    0.016841
3533    0.075714
3662    0.078830
3700    0.052831
3719    0.054382
3852    0.040986
3994    0.300996
Name: shot_statsbomb_xg, dtype: float64

Con 'location' obtendremos las coordenadas (x, y) de cada tiro.

In [ ]:
shots['location']

586     [101.0, 22.0]
647     [113.7, 39.9]
834      [95.6, 44.4]
1173     [94.0, 46.1]
1343    [104.1, 22.2]
1368    [113.0, 40.0]
1393     [93.7, 32.7]
1417    [109.8, 58.4]
1645    [107.5, 23.2]
1799    [106.0, 46.0]
1824    [101.0, 35.6]
1871    [110.8, 51.6]
1963    [111.7, 40.2]
2022    [109.1, 37.8]
2087    [112.9, 28.4]
2207    [100.4, 26.7]
2377    [100.9, 60.5]
2675    [117.4, 30.7]
2740     [98.7, 34.4]
2748    [106.1, 32.9]
2816    [109.8, 38.8]
3297    [107.4, 49.4]
3420     [94.4, 32.7]
3533    [111.1, 47.6]
3662    [102.6, 39.6]
3700    [102.5, 27.1]
3719    [103.4, 27.3]
3852     [95.6, 49.9]
3994    [113.0, 38.2]
Name: location, dtype: object

Con 'shot_end_location' veremos las coordenadas de donde finalizan los tiros.

In [ ]:
shots['shot_end_location']

586     [118.0, 38.0, 0.2]
647     [118.1, 44.9, 1.5]
834     [120.0, 36.8, 2.2]
1173    [119.1, 43.1, 1.5]
1343         [105.0, 23.2]
1368         [114.3, 40.0]
1393          [98.1, 33.3]
1417         [110.3, 57.4]
1645    [120.0, 46.0, 0.2]
1799    [117.8, 42.0, 0.2]
1824         [110.2, 39.3]
1871         [111.3, 51.0]
1963         [112.2, 40.4]
2022    [120.0, 37.4, 0.5]
2087    [120.0, 35.2, 2.4]
2207         [102.6, 27.9]
2377         [107.8, 52.5]
2675         [118.2, 61.7]
2740    [117.8, 37.1, 0.8]
2748    [120.0, 45.8, 0.1]
2816    [120.0, 36.8, 0.2]
3297    [116.6, 42.1, 0.2]
3420    [120.0, 39.8, 6.7]
3533         [113.2, 45.5]
3662    [120.0, 42.9, 2.0]
3700    [118.0, 40.6, 2.2]
3719    [120.0, 41.4, 2.4]
3852    [118.1, 42.6, 0.5]
3994    [120.0, 36.1, 0.1]
Name: shot_end_location, dtype: object

Accedemos a los valores de las coordenadas combinando  .values y los índices de las listas.

In [ ]:
shots['location'].values[0][0]

101.0

Esta es la coordenada y del tiro.

In [ ]:
shots['location'].values[0][1]

22.0